File: LeetCode/1004.Max_Consecutive_Ones_III/Mario/1004.Max_Consecutive_Ones_III.py
#!/usr/bin/env python3


class Solution:
    def longestOnes(self, nums: List[int], k: int) -> int:
        # sliding window,
        # l, r
        # sum(nums[l:r]) = how many 1s
        # sum + k >= (l -r + 1)
        # res = max(l - r + 1, res)

        # if sum + k > (l - r +1):
        # expand
        # if sum + k < (l - r + 1)
        # shrink
        l = count = 0

        res = 0
        for i, num in enumerate(nums):
            count += num
            # [l, i] zeros > ones, so have to shrink
            while count + k < (i - l + 1):
                count -= nums[l]
                l += 1
            res = max(res, i - l + 1)

        return res

File: LeetCode/1004.Max_Consecutive_Ones_III/1004.Max_Consecutive_Ones_III.swift
class Solution {
    func longestOnes(_ nums: [Int], _ k: Int) -> Int {
        // max num of consecutive 1's if flip at most k 0's
        // sliding window, count flip num of 0's
        // if > k, shift left pointer to right
        // if < k, shift right pointer

        let n = nums.count
        guard n > k else { return n }

        var l = 0, cnt = 0
        var res = 0

        for (i, num) in nums.enumerated() {
            cnt = num == 0 ? cnt + 1 : cnt

            while cnt > k && l <= i {
                cnt = nums[l] == 0 ? cnt - 1 : cnt
                l += 1
            }

            res = max(res, i - l + 1)
        }

        return res
    }
}
